Readme.md

# metrics-helpscout

A [Helpscout](https://www.helpscout.net/) plugin for [segmentio/metrics](https://github.com/segmentio/metrics).

Use this plugin to visualize your active ticket count, and who is behind on their tickets.

![](https://f.cloud.github.com/assets/658544/2361183/33c4df78-a62e-11e3-9921-6591e787e43e.png)

## Installation

    $ npm install metrics-helpscout 

## Example

```js
var Metrics = require('metrics');
var helpscout = require('metrics-helpscout');

new Metrics()
  .every('30m', helpscout('apiKey', ['mailbox-id']));
```

## Metrics

The metrics exposed by this plugin are:

Total active:
- `helpscout active tickets` - the total number of active tickets in teh mailboxes

Weekly:
- `helpscout tickets modified avg` - the trailing average of tickets modified in the last 7 days
- `helpscout tickets modified last week` - tickets modified in the last 7 days
- `helpscout tickets modified 2 weeks ago` - tickets modified 2 weeks ago
- `helpscout tickets created avg` - trailing average of tickets created in the last 7 days
- `helpscout tickets created last week` - tickets created in the last 7 days
- `helpscout tickets created 2 weeks ago` - tickets created two weeks ago

By Owner:
- `helpscout active tickets by owner` - active tickets by owner

```js
{
    "steve": 14,
    "mark": 3
}
```

Oldest Ticket:
- `helpscout oldest ticket time` - the javacript date of the oldest ticket
- `helpscout oldest ticket owner` - the owner of the oldest ticket
- `helpscout oldest ticket timeago` - the timeago string of the oldest ticket
- `helpscout oldest ticket shaming` - Example: "Bob: 2 days of no response"

Winners:
- `helpscout first place owner` - the person who closed the most tickets today
- `helpscout first place closed` - the winning number of tickets
- `helpscout second place owner` - the second place person
- `helpscout second place closed` - the second winning number of tickets
- `helpscout tickets closed today by owner` - tickets closed today by owner

```js
{
    "devin": 23,
    "randy": 15
}
```

## Quickstart

Here's a full example of a [Geckoboard](https://github.com/segmentio/geckoboard) dashboard showing support metrics:

```js
var Metrics = require('metrics');
var helpscout = require('metrics-helpscout');
var geckoboard = require('geckoboard')('api-key');

Metrics()
  .every('10m', helpscout('apiKey', 'mailbox-id'))
  .use(function (metrics) {
    metrics.on('helpscout tickets modified avg', geckoboard('widget-id').number);
});
```

## License

MIT

History.md

0.2.0 / 2014-03-31
==================



0.2.0 / 2014-03-31
==================

 * renaming metrics and adding created metrics
 * adding first and second place, changing oldest
0.1.0 - March 21, 2014
-------------------------
- adding first and second place winners
- renaming oldest ticket

0.0.1 - February 27, 2014
-------------------------
:sparkles:


index.js

var Batch = require('batch');
var Dates = require('date-math');
var debug = require('debug')('metrics:helpscout');
var Helpscout = require('helpscout');
var range = require('range');
var timeago = require('timeago');
timeago.settings.strings.suffixAgo = ''; // remove "ago"

/**
 * Expose `plugin`.
 */

module.exports = plugin;

/**
 * Create a new helpscout metrics plugin.
 *
 * @param {String} apiKey
 */

function plugin (apiKey, mailboxes) {
  return function helpscout (metrics) {
    var batch = new Batch();
    mailboxes.forEach(function (mailboxId) {
      batch.push(function (done) {
        var mailbox = Helpscout(apiKey, mailboxId);
        conversations(mailbox, done);
      });
    });
    debug('querying helpscout mailboxes %s ..', mailboxes);
    batch.end(function (err, mailboxes) {
      if (err) return debug('failed to query helpscout: %s', err);
      debug('finished querying helpscout mailboxes');
      var convos = []; // combine all the conversations
      mailboxes.forEach(function (c) { convos = convos.concat(c); });

      totalActive(metrics, convos);
      weekly(metrics, convos);
      oldestBreakdown(metrics, convos);
      todayBreakdown(metrics, convos);
    });
  };
}

/**
 * Calculate the total active tickets.
 *
 * @param {Array|Conversation} convos
 * @param {Object} results
 */

function totalActive (metrics, convos) {
  var active = convos.filter(function (convo) { return convo.status === 'active'; });
  metrics.set('helpscout active tickets', active.length);
}

/**
 * Calculate the total amount of tickets over the last two weeks.
 *
 * @param {Array|Conversation} convos
 * @param {Object} results
 */

function weekly (metrics, convos) {
  var now = new Date();
  var weekAgo = Dates.day.shift(now, -7);
  var twoWeeksAgo = Dates.day.shift(now, -14);

  var lastWeekModified = userModifiedAt(convos, weekAgo, now);

  metrics.set('helpscout tickets modified avg', Math.round(lastWeekModified.length / 7));
  metrics.set('helpscout tickets modified last week', lastWeekModified.length);
  metrics.set('helpscout tickets modified 2 weeks ago', userModifiedAt(convos, twoWeeksAgo, weekAgo).length);

  var lastWeekCreated = createdAt(convos, weekAgo, now);

  metrics.set('helpscout tickets created avg', Math.round(lastWeekCreated.length / 7));
  metrics.set('helpscout tickets created last week', lastWeekCreated.length);
  metrics.set('helpscout tickets created 2 weeks ago', createdAt(convos, twoWeeksAgo, weekAgo).length);
}

/**
 * Calculate oldest tickets by owner.
 *
 * @param {Array|Conversation} convos
 * @param {Object} results
 */

function oldestBreakdown (metrics, convos) {
  var breakdown = {};

  var oldestTime = null;
  var oldestOwner = null;

  var active = convos.filter(function (convo) { return convo.status === 'active'; });

  active.forEach(function (convo) {
    var owner = convo.owner;
    if (owner) {
      var n = name(owner);
      if (!breakdown[n]) breakdown[n] = 0;
      breakdown[n] += 1;

      var userModified = (new Date(convo.userModifiedAt)).getTime();
      // track oldest
      if (!oldestTime || oldestTime > userModified) {
        oldestTime = userModified;
        oldestOwner = n;
      }
    }
  });

  metrics.set('helpscout active tickets by owner', breakdown);

  metrics.set('helpscout oldest ticket time', new Date(oldestTime));
  metrics.set('helpscout oldest ticket owner', oldestOwner);
  metrics.set('helpscout oldest ticket timeago', timeago(new Date(oldestTime)));
  metrics.set('helpscout oldest ticket shaming', oldestOwner + ': ' + timeago(new Date(oldestTime)) + ' of no response.');
}

/**
 * Calculate today tickets by owner.
 *
 * @param {Array|Conversation} convos
 * @param {Object} results
 */

function todayBreakdown (metrics, convos) {
  var breakdown = {};

  var end = ceil(new Date());
  var start = floor(new Date());

  closedAt(convos, start, end).forEach(function (convo) {
    var owner = convo.owner;
    if (owner) {
      var n = name(owner);
      if (!breakdown[n]) breakdown[n] = 0;
      breakdown[n] += 1;
    }
  });

  var sorted = Object.keys(breakdown).sort(function (k1, k2) {
    return breakdown[k1] > breakdown[k2] ? -1: 1;
  });

  var first = sorted.length > 0 ? sorted[0] : null;
  var second = sorted.length > 1 ? sorted[1] : null;

  if (first) {
    metrics.set('helpscout first place owner', first);
    metrics.set('helpscout first place closed', breakdown[first]);
  }
  if (second) {
    metrics.set('helpscout second place owner', second);
    metrics.set('helpscout second place closed', breakdown[second]);
  }

  metrics.set('helpscout tickets closed today by owner', breakdown);
}


/**
 * Get a friendly name for a conversation assigned `owner`.
 *
 * @param {Owner} owner
 * @return {String}
 */

function name (owner) {
  return owner.firstName; // + ' ' + owner.lastName.charAt(0);
}

/**
 * Get all the conversations in a Helpscout `mailbox`.
 *
 * @param {Mailbox} mailbox
 * @param {Function} callback
 */

function conversations (mailbox, callback) {
  var convos = [];
  mailbox.conversations.list(function (err, res) {
    if (err) return callback(err);
    convos.push.apply(convos, res.items);
    if (res.page === res.pages) return callback(null, convos);
    var batch = new Batch();
    batch.concurrency(5);
    range(res.page+1, res.pages+1).forEach(function (page) {
      batch.push(function (done) {
        debug('fetching conversations page %d / %d ..', page, res.pages);
        mailbox.conversations.list({ page: page }, done);
      });
    });
    batch.end(function (err, responses) {
      if (err) {
        debug('helpscout error: %s', err.toString());
        return callback(err);
      }
      debug('fetched all conversation pages');
      responses.forEach(function (res) { convos.push.apply(convos, res.items); });
      callback(null, convos);
    });
  });
}

/**
 * Filter the helpscout `convos` by `start` and `end` createdAt date.
 *
 * @param {Array|Conversation} convos
 * @param {Date} start
 * @param {Date} end
 * @return {Array|Conversation}
 *
 */
function createdAt (convos, start, end) {
  var s = start.getTime();
  var e = end.getTime();
  return convos.filter(function (convo) {
    var createdAt = (new Date(convo.createdAt)).getTime();
    return createdAt >= s && createdAt <= e;
  });
}

/**
 * Filter the helpscout `convos` by `start` and `end` userModifiedAt date.
 *
 * @param {Array|Conversation} convos
 * @param {Date} start
 * @param {Date} end
 * @return {Array|Conversation}
 *
 */
function userModifiedAt (convos, start, end) {
  var s = start.getTime();
  var e = end.getTime();
  return convos.filter(function (convo) {
    var userModifiedAt = (new Date(convo.userModifiedAt)).getTime();
    return userModifiedAt >= s && userModifiedAt <= e;
  });
}

/**
 * Filter the helpscout `convos` by `start` and `end` closedAt date.
 *
 * @param {Array|Conversation} convos
 * @param {Date} start
 * @param {Date} end
 * @return {Array|Conversation}
 *
 */
function closedAt (convos, start, end) {
  var s = start.getTime();
  var e = end.getTime();
  return convos.filter(function (convo) {
    if (!convo.closedAt) return false;
    var c = (new Date(convo.closedAt)).getTime();
    return c >= s && c <= e;
  });
}


/**
 * Floor the `date` to the nearest day,
 * while keeping in the same locale
 * (unlike UTC'ing like Dates.day.floor).
 */

function floor (date) {
  date = new Date(date);
  date.setHours(0);
  date.setMinutes(0);
  date.setSeconds(0);
  return date;
}

/**
 * Floor the `date` to the nearest day,
 * while keeping in the same locale
 * (unlike UTC'ing like Dates.day.floor).
 */

function ceil (date) {
  date = new Date(date);
  date.setHours(23);
  date.setMinutes(59);
  date.setSeconds(59);
  return date;
}